### PR TITLE
Add TimeInForce.Day support in backtesting and IB brokerage

### PR DESCRIPTION
**Pull request by StefanoRaggi**
#### Description
This PR includes some common changes to support `TimeInForce` types:
- New `ITimeInForceHandler` interface and `GoodTilCancelled` and `Day` implementations
- Base `OrderProperties` class, instantiated in `QCAlgorithm` constructor

Both unit tests and regression algorithms have been added.

This feature is currently supported only in backtesting and with Interactive Brokers brokerage.

#### Related Issue
#1093 

#### Motivation and Context
Previously we only supported `GTC` time in force.

#### Requires Documentation Change
No.

#### How Has This Been Tested?
New unit tests, new regression test and IB live testing.

#### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/ect)

#### Checklist:
- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm.Python/TimeInForceAlgorithm.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+from QuantConnect.Algorithm import *
+from QuantConnect.Orders import *
+
+### <summary>
+### Demonstration algorithm of time in force order settings.
+### </summary>
+### <meta name="tag" content="using data" />
+### <meta name="tag" content="using quantconnect" />
+### <meta name="tag" content="trading and orders" />
+class TimeInForceAlgorithm(QCAlgorithm):
```
Comment (mchandschuh) [inline Algorithm.Python/TimeInForceAlgorithm.py]: Thanks for adding a python version too!

File: Brokerages/Backtesting/BacktestingBrokerage.cs [hunk 2]
```diff
@@ -457,5 +480,27 @@ private void SetPendingOrder(Order order)
             // only save off clones!
             _pending[order.Id] = order.Clone();
         }
+
+        private ITimeInForceHandler GetTimeInForceHandler(TimeInForce timeInForce)
```
Comment (mchandschuh) [inline Brokerages/Backtesting/BacktestingBrokerage.cs]: instead of having this function, it might be more concise to simply initialize the dictionaries with all the handlers already mapped, and then replace calls to this function with explicit dictionary access, `_timeInForceHandlers[timeInForce]`

File: Common/Orders/TimeInForces/DayTimeInForceHandler.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        /// <summary>
+        /// Checks if an order has expired
+        /// </summary>
+        /// <param name="order">The order to be checked</param>
+        /// <returns>Returns true if the order has expired, false otherwise</returns>
+        public bool HasOrderExpired(Order order)
+        {
+            var exchangeHours = MarketHoursDatabase
+                .FromDataFolder()
+                .GetExchangeHours(order.Symbol.ID.Market, order.Symbol, order.Symbol.SecurityType);
```
Comment (mchandschuh) [inline Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: Instead of accessing the `FromDataFolder()` here, I think it may be better to accept the market hours database instance in the constructor, and then who ever creates this type can provide their own market hours -- we'll pretty much always use it from the data folder, but this can make this class testable w/out having a `market-hours-database.json` file in the correct place.

File: Common/Orders/TimeInForces/DayTimeInForceHandler.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+            var time = _algorithm.UtcTime.ConvertFromUtc(exchangeHours.TimeZone);
+
+            return !exchangeHours.IsOpen(time, false);
```
Comment (mchandschuh) [inline Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: Is this correct? Isn't it possible to move to the next day and have `IsOpen` return true again. Is it more correct to compare the order's submit date and the current date (in the security's time zone) ?

What happens with 24 hour markets? If I say time in force = day at 23:59PM - does it cancel at the stroke of midnight?

File: Tests/Common/Orders/TimeInForces/TimeInForceHandlerTests.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+            var fill1 = new OrderEvent(order.Id, order.Symbol, DateTime.UtcNow, OrderStatus.PartiallyFilled, OrderDirection.Buy, order.LimitPrice, 3, 0);
+            Assert.IsTrue(handler.IsFillValid(order, fill1));
+
+            var fill2 = new OrderEvent(order.Id, order.Symbol, DateTime.UtcNow, OrderStatus.Filled, OrderDirection.Buy, order.LimitPrice, 7, 0);
+            Assert.IsTrue(handler.IsFillValid(order, fill2));
+        }
+
+        [Test]
+        public void DayTimeInForceOrderExpiresAtMarketClose()
```
Comment (mchandschuh) [inline Tests/Common/Orders/TimeInForces/TimeInForceHandlerTests.cs]: This assumes a couple things -- one, that we get data at market close (almost guaranteed) or that we get data before the next market open (less of a guarantee) -- see comments in the day handler impl.

File: Common/Orders/TimeInForces/DayTimeInForceHandler.cs (hunk 4 above)
Comment (StefanoRaggi) [reply Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: Good point, this might need to be different per security type:
- Forex: both FXCM and Oanda use 5 PM New York time as day close
- Crypto: GDAX does not support Day, so maybe we should use midnight UTC here?

File: Common/Orders/TimeInForces/DayTimeInForceHandler.cs [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        /// Checks if an order has expired
+        /// </summary>
+        /// <param name="order">The order to be checked</param>
+        /// <returns>Returns true if the order has expired, false otherwise</returns>
+        public bool HasOrderExpired(Order order)
+        {
+            var exchangeHours = _marketHoursDatabase.GetExchangeHours(order.Symbol.ID.Market, order.Symbol, order.Symbol.SecurityType);
+
+            var orderTime = order.Time.ConvertFromUtc(exchangeHours.TimeZone);
+            var time = _algorithm.UtcTime.ConvertFromUtc(exchangeHours.TimeZone);
```
Comment (mchandschuh) [inline Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: This algorithm dependency seems heavy handed. It looks like we're only using the algorithm to get the current time. A common pattern in lean models is to accept the `Security` object in the method. Accepting the security object would remove the algorithm dependency *and* the market hours database dependency, since the current time and the exchange hours can both be directly resolved form the security object.

File: Algorithm.CSharp/TimeInForceAlgorithm.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        /// </summary>
+        /// <param name="data">Slice object keyed by symbol containing the stock data</param>
+        public override void OnData(Slice data)
+        {
+            if (_gtcOrderTicket == null)
+            {
+                // This order has a default time in force of GoodTilCancelled,
+                // it will never expire and will not be cancelled automatically.
+
+                _gtcOrderTicket = LimitOrder(_symbol, 10, 160m);
```
Comment (mchandschuh) [inline Algorithm.CSharp/TimeInForceAlgorithm.cs]: For consistency, perhaps set `DefaultOrderProperties.TimeInForce = TimeInforce.GoodTilCancelled`

File: Brokerages/Backtesting/BacktestingBrokerage.cs [hunk 8]
```diff
@@ -38,6 +39,7 @@ public class BacktestingBrokerage : Brokerage
         private readonly ConcurrentDictionary<int, Order> _pending;
         private readonly object _needsScanLock = new object();
         private readonly HashSet<Symbol> _pendingOptionAssignments = new HashSet<Symbol>();
+        private readonly Dictionary<TimeInForce, ITimeInForceHandler> _timeInForceHandlers = new Dictionary<TimeInForce, ITimeInForceHandler>();
```
Comment (mchandschuh) [inline Brokerages/Backtesting/BacktestingBrokerage.cs]: Hopefully we can change the implementations of `ITimeInForceHandler` such that they pull all information from the method parameters leading to a parameterless constructor, and then we can inline the creating of this map here of the form:
``` csharp
private readonly Dictionary<TimeInForce, ITimeInForceHandler> _timeInForceHandlers = new Dictionary<TimeInForce, ITimeInForceHandler>
{
    { TimeInForce.GoodTilCanceled, new GoodTilCanceledTimeInForceHandler() }
};
```

File: Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs [hunk 9]
```diff
@@ -1898,6 +1891,66 @@ private static OrderType ConvertOrderType(IBApi.Order order)
...
+
+                //case TimeInForce.FillOrKill:
+                //    return IB.TimeInForce.FillOrKill;
+
+                //case TimeInForce.ImmediateOrCancel:
+                //    return IB.TimeInForce.ImmediateOrCancel;
+
+                case TimeInForce.GoodTilCancelled:
+                default:
+                    return IB.TimeInForce.GoodTillCancel;
```
Comment (mchandschuh) [inline Brokerages/InteractiveBrokers/InteractiveBrokersBrokerage.cs]: Something I'm just realizing. In english 'canceled' and 'cancelled' are both equally correct spellings. We already use 'canceled' with a single L in `OrderStatus.Canceled` -- so let's keep the spelling the same for all our sanity :P

File: Common/Brokerages/FxcmBrokerageModel.cs [hunk 10]
```diff
@@ -122,6 +122,16 @@ public override bool CanSubmitOrder(Security security, Order order, out Brokerag
                 return IsValidOrderPrices(security, OrderType.StopLimit, stopLimit.Direction, stopLimit.StopPrice, stopLimit.LimitPrice, ref message);
             }
 
+            // validate time in force
+            if (order.TimeInForce != TimeInForce.GoodTilCancelled)
+            {
+                message = new BrokerageMessageEvent(BrokerageMessageType.Warning, "NotSupported",
+                    "This model does not support " + order.TimeInForce + " time in force."
```
Comment (mchandschuh) [inline Common/Brokerages/FxcmBrokerageModel.cs]: awesomely perfect and simple, great stuff!

File: Common/Brokerages/InteractiveBrokersBrokerageModel.cs [hunk 11]
```diff
@@ -62,24 +62,38 @@ public override bool CanSubmitOrder(Security security, Order order, out Brokerag
         {
             message = null;
 
+            // validate security type
+            if (security.Type != SecurityType.Equity &&
+                security.Type != SecurityType.Forex &&
+                security.Type != SecurityType.Option &&
+                security.Type != SecurityType.Future)
```
Comment (mchandschuh) [inline Common/Brokerages/InteractiveBrokersBrokerageModel.cs]: no change required, but typically when there's a set we're comparing against, it's often cleaner to have:
``` csharp
private static readonly HashSet<SecurityType> SupportedSecurityTypes = new HashSet<SecurityType>{ SecurityType.Equity, SecurityType.Forex, SecurityType.Option, SecurityType.Future };

if (!SupportedSecurityTypes.Contains(security.Type)
{
    // error
}
```

File: Common/Brokerages/InteractiveBrokersBrokerageModel.cs [hunk 12]
```diff
@@ -62,24 +62,38 @@ public override bool CanSubmitOrder(Security security, Order order, out Brokerag
...
             //https://www.interactivebrokers.com/en/?f=%2Fen%2Ftrading%2FforexOrderSize.php
-            switch (order.SecurityType)
+            if (security.Type == SecurityType.Forex &&
+                !IsForexWithinOrderSizeLimits(order.Symbol.Value, order.Quantity, out message))
+            {
+                return false;
+            }
+
+            // validate time in force
+            if (order.TimeInForce != TimeInForce.GoodTilCancelled && order.TimeInForce != TimeInForce.Day)
```
Comment (mchandschuh) [inline Common/Brokerages/InteractiveBrokersBrokerageModel.cs]: Similar comment to the above, often cleaner to define a collection of supported cases and check the collection. This makes the code read nicer and makes it easier to update the conditions in the future as we add/remove support for various things.

File: Common/Interfaces/IAlgorithm.cs [hunk 13]
```diff
@@ -328,6 +328,11 @@ IFutureChainProvider FutureChainProvider
             get;
         }
 
+        /// <summary>
+        /// Gets the market hours database in use by this algorithm
+        /// </summary>
+        MarketHoursDatabase MarketHoursDatabase { get; }
```
Comment (mchandschuh) [inline Common/Interfaces/IAlgorithm.cs]: I'm reluctant about this. In theory, this is a great idea, but I have concerns on the direction of the dependency. Currently QCAlgorithm is hard-coded to fetch `FromDataFolder()` (as is the rest of the application). I love the intent behind this, making things instance based, but if we're going to make accesses to the market hours db instance based we should tackle that in a separate PR and update the solution to follow that new pattern.

File: Common/Interfaces/ITimeInForceHandler.cs [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    /// Handles the time in force for an order
+    /// </summary>
+    public interface ITimeInForceHandler
+    {
+        /// <summary>
+        /// Checks if an order has expired
+        /// </summary>
+        /// <param name="order">The order to be checked</param>
+        /// <returns>Returns true if the order has expired, false otherwise</returns>
+        bool HasOrderExpired(Order order);
```
Comment (mchandschuh) [inline Common/Interfaces/ITimeInForceHandler.cs]: We spoke about adding the `Security` instance here to match existing patterns w.r.t models in lean. For example, look at the fee/fill/slippage models, all of which accept the security object in addition to the order object. This will remove the need for the algorithm and market hours db dependencies

File: Algorithm/QCAlgorithm.cs [hunk 15]
```diff
@@ -72,7 +72,7 @@ public partial class QCAlgorithm : MarshalByRefObject, IAlgorithm
         /// <summary>
         /// Gets the market hours database in use by this algorithm
         /// </summary>
-        protected MarketHoursDatabase MarketHoursDatabase { get; }
+        public MarketHoursDatabase MarketHoursDatabase { get; }
```
Comment (mchandschuh) [inline Algorithm/QCAlgorithm.cs]: Please revert this change.

File: Brokerages/Backtesting/BacktestingBrokerage.cs [hunk 16]
```diff
@@ -361,6 +385,12 @@ public void Scan()
 
                     foreach (var fill in fills)
                     {
+                        // check if the fill should be emitted
+                        if (!timeInForceHandler.IsFillValid(security, order, fill))
+                        {
+                            break;
```
Comment (mchandschuh) [inline Brokerages/Backtesting/BacktestingBrokerage.cs]: OK for now, but this will change, we can't assume the implementation of `IsFillValid` negates all fills in the event there are multiple fills -- when we wire up order event projections this will change, so it's ok for now.

File: Common/Orders/TimeInForces/DayTimeInForceHandler.cs [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+            var orderTime = order.Time.ConvertFromUtc(exchangeHours.TimeZone);
+            var time = security.LocalTime;
+
+            bool expired;
+            switch (order.SecurityType)
+            {
+                case SecurityType.Forex:
+                case SecurityType.Cfd:
+                    // expires at 5 PM NewYork time
+                    var expiryTime = new DateTime(orderTime.Date.Year, orderTime.Date.Month, orderTime.Date.Day, 17, 0, 0);
```
Comment (mchandschuh) [inline Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: we say they expire at 5PM new york but are converting to exchange tz with no guarantee that it's new york time. Is the 5PM represented in the market hours segments? Maybe there's a more general way to define this cut off time? I don't know the answer, but worth considering.

File: Common/Orders/TimeInForces/DayTimeInForceHandler.cs [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+                    expired = time.Date > orderTime.Date || time.ConvertTo(exchangeHours.TimeZone, TimeZones.NewYork) >= expiryTime;
```
Comment (mchandschuh) [inline Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: What if I submit an order at 5:01PM -- is that order good until the next day's 5PM, in which case I think the first part of this condition is incorrect.

File: Brokerages/Backtesting/BacktestingBrokerage.cs [hunk 19]
```diff
@@ -39,6 +40,14 @@ public class BacktestingBrokerage : Brokerage
         private readonly object _needsScanLock = new object();
         private readonly HashSet<Symbol> _pendingOptionAssignments = new HashSet<Symbol>();
 
+        private readonly Dictionary<TimeInForce, ITimeInForceHandler> _timeInForceHandlers = new Dictionary<TimeInForce, ITimeInForceHandler>
+        {
+            { TimeInForce.GoodTilCanceled, new GoodTilCanceledTimeInForceHandler() },
+            { TimeInForce.Day, new DayTimeInForceHandler() },
+            // Custom time in force will be renamed to GTD soon and will have its own new handler
+            { TimeInForce.Custom, new GoodTilCanceledTimeInForceHandler() }
```
Comment (mchandschuh) [inline Brokerages/Backtesting/BacktestingBrokerage.cs]: This looks nice and clean! With the upcoming refactor this is going to change (again) such that the implementation will ride along w/ the order object -- but for now it's nice a clear what the mappings are.

File: Common/Orders/TimeInForces/DayTimeInForceHandler.cs [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+
+                    orderTime = order.Time.ConvertFromUtc(cutOffTimeZone);
+                    var expiryTime = orderTime.Date.Add(cutOffTimeSpan);
+                    if (orderTime.TimeOfDay > cutOffTimeSpan)
+                    {
+                        // order submitted after 5 PM, expiry on next date
+                        expiryTime = expiryTime.AddDays(1);
+                    }
+
+                    expired = time.ConvertTo(exchangeHours.TimeZone, cutOffTimeZone) >= expiryTime;
```
Comment (mchandschuh) [inline Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: as a reminder, we discussed using `exchangeHours.GetNextMarketClose(orderTime, false)` to determine when the end of 'day' should be.

Comment (mchandschuh) [reply Common/Orders/TimeInForces/DayTimeInForceHandler.cs]: you're 100% correct, it won't work that way -- let's keep it as you have it. If we ever need a more dynamic value we can look into sourcing the value through the brokerage model.
